src/home/user-record-util.c

```diff
@@ -1404,6 +1404,210 @@ int user_record_is_supported(UserRecord *hr, sd_bus_error *error) {
+                        if (r < 0)
+                                return r;
+                        if (r == 0) {
+                                /* It's only safe to change anything inside of matching perMachine sections */
+                                r = json_variant_append_array(&new_per_machine, e);
+                                if (r < 0)
+                                        return r;
+                                continue;
+                        }
+
+                        z = json_variant_ref(e);
+
+                        r = remove_safe_json_fields_regular(&z);
+                        if (r < 0)
+                                return r;
+
+                        /* Note that the resulting object will never be empty, because it will have at
+                         * least matchMachineId and/or matchHostname. We also deem these unsafe to change,
+                         * since they can drastically change the behavior of this user on other machines */
+
+                        r = json_variant_append_array(&new_per_machine, z);
+                        if (r < 0)
+                                return r;
+                }
+
+                if (json_variant_is_blank_array(new_per_machine))
+                        r = json_variant_filter(v, STRV_MAKE("perMachine"));
+                else
+                        r = json_variant_set_field(v, "perMachine", new_per_machine);
+                if (r < 0)
+                        return r;
+        }
+
+        /* Handle the privileged section */
+        privileged = json_variant_ref(json_variant_by_key(*v, "privileged"));
+        if (privileged) {
+                r = json_variant_filter(&privileged, safe_privileged);
+                if (r < 0)
+                        return r;
+
+                if (json_variant_is_blank_object(privileged))
+                        r = json_variant_filter(v, STRV_MAKE("privileged"));
+                else
+                        r = json_variant_set_field(v, "privileged", privileged);
+                if (r < 0)
+                        return r;
+        }
+
+        return 0;
+}
+
+int user_record_changes_are_safe(UserRecord *a, UserRecord *b) {
+        _cleanup_(json_variant_unrefp) JsonVariant *va = NULL, *vb = NULL;
+        int r;
+
+        if (!a && !b)
+                return true;
+        if (!a || !b)
+                return false;
+
+        r = user_group_record_mangle(a->json, USER_RECORD_EXTRACT_SIGNABLE|USER_RECORD_PERMISSIVE, &va, NULL);
+        if (r < 0)
+                return r;
+
+        r = remove_safe_json_fields(&va);
+        if (r < 0)
+                return r;
+
+        r = user_group_record_mangle(b->json, USER_RECORD_EXTRACT_SIGNABLE|USER_RECORD_PERMISSIVE, &vb, NULL);
+        if (r < 0)
+                return r;
+
+        r = remove_safe_json_fields(&vb);
+        if (r < 0)
+                return r;
+
+        return json_variant_equal(va, vb);
+}
```

Comment:
would be good to have a test cse for this btw, this is security sensitive after all, would be good to know that it actually works.

Reply:
Added